### PR TITLE
Improve EverCrypt Hash API and fix #598, #728, #729

### LINKED ISSUE #598: Make monomorphizations in EverCrypt.DRBG private
No reason to keep them. Agile API is the way.

### LINKED ISSUE #728: Tweak bundling to hide EverCrypt.Hash underneath EverCrypt.Hash.Incremental
**Describe the bug**

This is mostly a code-quality issue. Just for SHA2, we would like to only expose one-shot and incremental APIs for EverCrypt.Hash. The way to go is to:
- move the agile one-shot hash to incremental
- change E_HASH_BUNDLE in the Makefile to be `-bundle EverCrypt.Hash.Incremental=EverCrypt.Hash`
- maybe provide a type monomorphization naming hint for the evecrypt incremental hash state type
- hopefully this will eliminate all of the uu__'s in the header file

### PR DESCRIPTION
**Pull request by msprotz**
## Proposed changes

This is a code-quality issue, and hides the incremental evercrypt hash API in favor of just the one-shot and streaming ones. This is in line with my earlier change for SHA2, and fixes #728, but also #599 (to some extent).

Look at the headers for the changes.

## Types of changes

What types of changes does your code introduce to HACL*?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CODE_OF_CONDUCT.md](CODE_OF_CONDUCT.md) doc
- [x] I have read and agree to submit my changes under the [LICENSE](LICENSE.md)
- [x] I have added necessary documentation (if appropriate)
- [x] I have edited [CHANGES.md](CHANGES.md) (if appropriate)

### CODE REVIEW

**Comment by msprotz** [comment]
Overall, this removes quite a large amount of C code, which naturally puts me in quite a good mood.

File: dist/gcc-compatible/EverCrypt_Hash.h [hunk 1]
```diff
@@ -47,236 +47,47 @@ extern "C" {
...
-EverCrypt_Hash_Incremental_finish_blake2b(
-  Hacl_Streaming_Functor_state_s___EverCrypt_Hash_state_s____ *p,
-  uint8_t *dst
-);
-
-Spec_Hash_Definitions_hash_alg
-EverCrypt_Hash_Incremental_alg_of_state(
-  Hacl_Streaming_Functor_state_s___EverCrypt_Hash_state_s____ *s
-);
+void EverCrypt_Hash_Incremental_hash_224(uint8_t *input, uint32_t input_len, uint8_t *dst);
```
Comment (polubelova) [inline dist/gcc-compatible/EverCrypt_Hash.h:81]: is there a difference between `EverCrypt_Hash_Incremental_hash_224` and `EverCrypt_Hash_Incremental_hash`? if so, can we unify the order of arguments and add documentation?

**Comment by msprotz** [comment]
@polubelova I addressed your review comments, please take a look

@victor-dumitrescu I had to fix up a few things in the OCaml API... notably, a change I did while I was touching these files is that I removed the "specialized" APIs (e.g. evercrypt/hmac/sha256), on the basis that clients can always go through the agile API, and skipping one run-time test is not worth polluting the API and confusing the client as to which API they should go through

let me know if this is a problem for you? thanks!

**Comment by msprotz** [comment]
(The issue for hiding the monomorphizations is #598)

**Comment by victor-dumitrescu** [comment]
LGTM, I don't think the removals from the OCaml API pose any problem

**Comment by victor-dumitrescu** [comment]
Just a note that we should probably bump the opam version from 0.6.x to 0.7.x when this is merged

**Review by R1kM — APPROVED** [review]

File: providers/evercrypt/fst/EverCrypt.HMAC.fst [hunk 2]
```diff
@@ -3,6 +3,55 @@ module EverCrypt.HMAC
...
 open Hacl.HMAC
 
+// EverCrypt.Hash.Incremental.hash_256 is marked as private, so it is
+// inaccessible from here. Furthermore, the EverCrypt.Hash.Incremental module
+// does not have an interface, meaning that we can't even friend it! Writing an
+// interface for EverCrypt.Hash.Incremental is possible, but doesn't make much
+// sense: instantiations of the functor offer, by construction, an abstract
+// interface. Fortunately, we can use tactics to stealthily gain access to a
+// private definition that F* normally won't let us access.
+let super_hack () =
```
Comment (polubelova) [inline providers/evercrypt/fst/EverCrypt.HMAC.fst:13]: It would be nice to mention this hack in one of the tutorials (F* or Low*) :)

**Review by polubelova — APPROVED** [review]
Thanks!  [EverCrypt Hash API](https://github.com/hacl-star/hacl-star/blob/5bdfabbae6c0523eca7c241cb034b3bf87ea6998/dist/gcc-compatible/EverCrypt_Hash.h) looks good to me! Are you planning to document it?

What about HACL Hash API? [Hacl_Streaming_SHA2.h](https://github.com/hacl-star/hacl-star/blob/5bdfabbae6c0523eca7c241cb034b3bf87ea6998/dist/gcc-compatible/Hacl_Streaming_SHA2.h) and [Hacl_Hash_SHA2.h](https://github.com/hacl-star/hacl-star/blob/5bdfabbae6c0523eca7c241cb034b3bf87ea6998/dist/gcc-compatible/Hacl_Hash_SHA2.h) will be fixed in a follow-up PR?

File: providers/evercrypt/fst/EverCrypt.HMAC.fst (hunk 2 above)
Comment (msprotz) [reply providers/evercrypt/fst/EverCrypt.HMAC.fst:13]: But then too many people might become aware of it, and it might get removed... tricky!!

**Comment by msprotz** [comment]
- Documentation: ok I can push an extra commit to document the API in the style of others
- SHA2 and others: yes, in a follow-up PR, I need to fix #756 before I can do the same thing for SHA2.

Thanks for the review, and happy new year!

**Pull request merged**